src/Orleans.Core.Abstractions/Streams/Core/IAsyncBatchObserver.cs

```diff
@@ -0,0 +1,71 @@
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Orleans.Streams
+{
+    public class OrderedItem<T>
+    {
+        public OrderedItem(T item, StreamSequenceToken token)
+        {
+            this.Item = item;
+            this.Token = token;
+        }
+
+        public T Item { get; }
+        public StreamSequenceToken Token { get; }
+    }
+
+    /// <summary>
+    /// This interface generalizes the IAsyncObserver interface to allow production and consumption of batches of items.
+    /// <para>
+    /// Note that this interface is implemented by item consumers and invoked (used) by item producers.
+    /// This means that the consumer endpoint of a stream implements this interface.
+    /// </para>
+    /// </summary>
+    /// <typeparam name="T">The type of object consumed by the observer.</typeparam>
+    public interface IAsyncBatchObserver<T>
```

Comment:
Can we do something like following, which is in the code base now. 

```
public interface IAsyncBatchObserver<in T> : IAsyncObserver<T>{
Task OnNextAsync(IList<OrderedItem<T>> items)
}
```

So we can avoid having two ResumeAsync methods, and also two observers side by side in our StreamConsumer class, and other stream classes. 

And when delivering, if the observer implemeted IAsyncBatchObserver, then we call its OnBatchNext method, otherwise, we call the OnNext method. 

I'm not saying we should do that, since I haven't coded that direction myself so I don't know the possible naunce.  But this direction seems to be able to eliminate the need to have two observer side by side in our stream classes, which isn't ideal for readablity and maintenance.

Reply:
We can do this, but it's, IMO, a really bad public surface.

Implementors of IAsyncBatchObserver would need implement the non-batch onNextAsync, when its unnecessary.
It would be unclear to users under which conditions each onNextAsync was called.
It would not be clear that one should get a different behavior by subscribing using a batched observer, for either users or implementors of the observable.

Over all, I think we'd be adopting a bad public surface for the sake of avoiding making our already debt-ridden implementation mildly more debt-ridden.